c++/src/kj/CMakeLists.txt

```diff
@@ -137,6 +137,26 @@ if(NOT CAPNP_LITE)
   add_library(kj-async ${kj-async_sources})
   add_library(CapnProto::kj-async ALIAS kj-async)
   target_link_libraries(kj-async PUBLIC kj)
+  if (KJ_USE_FIBERS)
```

Comment:
I'll need @harrishancock's help to review cmake stuff...

Reply:
Sure. Note that I could move the `PkgConfig` part out into a cmake module, and just do `find_package(ucontext REQUIRED)` instead, but that would only earn 2 lines here (at the cost of adding a file). It's just in case you don't want to see pkg-config there.

In this case it felt more expressive like this, but I'm happy to change.

Reply:
Allow me to share my opinion but please note that I'm not a capnproto/kj maintainer, just have some opinions on CMake stuff.

In the top-level CMakeLists, I would handle the dependencies like `WITH_OPENSSL`, with `ON`, `OFF` and `AUTO` (default) options, and maybe the same naming scheme `WITH_FIBERS`?
And like `WITH_OPENSSL`, do the feature detection after the option.

https://github.com/capnproto/capnproto/blob/82b29296f76c583856511f645d9ab4e427a6d218/c%2B%2B/CMakeLists.txt#L44-L65

Then in kj's CMakeLists.txt, something like `WITH_OPENSSL`:

```cmake
  if (WITH_FIBERS)
    target_compile_definitions(kj-async PRIVATE KJ_USE_FIBERS)
    if (libucontext_FOUND)
      target_link_libraries(kj-async PUBLIC PkgConfig::libucontext)
    else()
  else()
    target_compile_definitions(kj-async PRIVATE KJ_USE_FIBERS=0)
  endif()
```

Then for CMake installation, propagate the libucontext dependencies like `WITH_OPENSSL`:

https://github.com/capnproto/capnproto/blob/82b29296f76c583856511f645d9ab4e427a6d218/c%2B%2B/cmake/CapnProtoConfig.cmake.in#L52-L63


Regarding `KJ_USE_FIBERS` being `PRIVATE`, I'm not sure it's correct:

```cmake
target_compile_definitions(kj-async PRIVATE KJ_USE_FIBERS=0) 
```

kj/async.h check/define it in its header, so I think it's part of the public API, and maybe should be propagated to consumers.
At least, the unit test file, should define it like `WITH_OPENSSL` here:

https://github.com/capnproto/capnproto/blob/82b29296f76c583856511f645d9ab4e427a6d218/c%2B%2B/src/kj/CMakeLists.txt#L273-L277

Note that using `set_source_files_properties` is wrong in the `WITH_OPENSSL` case, it would be better to `APPEND` the property with something like:

```cmake
  if (WITH_FIBERS)
    set_property(
      SOURCE async-test.c++
      APPEND PROPERTY COMPILE_DEFINITIONS KJ_USE_FIBERS
  )
  endif()
```

Reply:
> Regarding KJ_USE_FIBERS being PRIVATE, I'm not sure it's correct

Right, I honestly don't know. @harrishancock any opinion?

@Sarcasm: I get your opinion about the options, but to me it feels like it's not in the scope of this PR. This PR is about adding support for musl, not improving the options, right? Just to keep it contained :innocent:

Reply:
> This PR is about adding support for musl, not improving the options, right?

This PR is adding a CMake option, CMake options defined by the project is a sort of public API that packagers relies on.
So, it's nice to provide stability here and IMHO `WITH_FIBERS` is a better name for 2 reasons:

1. The corresponding autotools option is `--with-fibers` (also, there is `--with-openssl`).
2. `WITH_OPENSSL` is already named like this.

```
./configure --with-openssl --with-fibers
cmake -DWITH_OPENSSL=ON -DWITH_FIBERS=ON
# VS
cmake -DWITH_OPENSSL=ON -DKJ_USE_FIBERS=ON
```

If supporting `AUTO` if too annoying for you, that can be added later on in a retro-compatible way.
This "How to" can be of help too:
- https://sarcasm.github.io/notes/dev/cmake.html#optional-dependencies
  The example, like this PR, uses `pkg_check_modules()`.
  Disclaimer: I'm the author.

The `find_dependency()` and `PUBLIC/PRIVATE` stuff seems critical to me, let's see what Harris has to say.

I don't my trying to implement my suggestions if you prefer but I don't have the environment to test for `libucontext`.

Reply:
No sure, I can do it, I was not aware that `AUTO` was a thing in cmake, but you convinced me :blush:.

Reply:
@Sarcasm: I think I implemented all your suggestions but the `find_dependency` one. I'm not completely sure how to do it, because `libucontext` does not provide a cmake module (that's why I use pkg-config). So `find_dependency()` (which is a wrapper around `find_package`) won't "just find it", and I'm not sure if calling pkg-config will do exactly what's needed, will it?

Any opinion there?